### PR TITLE
Worked on Spotify artists, explored the results of the API and handle…

### PR DESCRIPTION
**Pull request by abaddonpuff**
Worked on Spotify artists, explored the results of the API and handled some pending TODOs on the X API.

### CODE REVIEW

File: webCommander/centralHub/models.py [hunk 1]
```diff
@@ -27,4 +27,17 @@ class Meta:
         unique_together = [('twitter_user','twitter_text')]
 
     def __str__(self):
-        return f'{self.twitter_user.twitter_handle} Posts'
\ No newline at end of file
+        return f'{self.twitter_user.twitter_handle} Posts'
+
+class SpotifyArtistInfo(models.Model):
```
Comment (bbelderbos) [inline webCommander/centralHub/models.py:32]: As discussed, let's make this a 1-to-many: artist->albums

File: webCommander/centralHub/spotify/spotifyAPI.py [hunk 2]
```diff
@@ -48,35 +51,56 @@ def get_spotify_artist(artist_name):
     Authenticate then search an artist based on an artist name. Searches for the name if id is not provided
     '''
 
-    if len(artist_name) != 22:
-        artist_id = search_spotify_artist(artist_name)
+    #TODO 
+    #Spotify artist is a 22 fixed character ID, but searching for a word would require to first call search_spotify_artist
```
Comment (bbelderbos) [inline webCommander/centralHub/spotify/spotifyAPI.py]: is this still needed after our call? I mean I saw type=artist in the api response or was that for another call?

File: webCommander/centralHub/spotify/spotifyAPI.py [hunk 3]
```diff
@@ -48,35 +51,56 @@ def get_spotify_artist(artist_name):
...
 
-    if len(artist_name) != 22:
-        artist_id = search_spotify_artist(artist_name)
+    #TODO 
+    #Spotify artist is a 22 fixed character ID, but searching for a word would require to first call search_spotify_artist 
+    #to get the id which would also return some ids even if not an artist. Not sure how to handle this.
 
-    headers = spotify_authenticate()
+    if len(artist_name) != SPOTIFY_FIXED_ARTIST_CHAR_LEN:
+        artist_id = search_spotify_artist(artist_name)
```
Comment (bbelderbos) [inline webCommander/centralHub/spotify/spotifyAPI.py]: I think we need to move this out of this function to first have the user choose the row / id, then call `get_spotify_artist` for that specific one

File: webCommander/centralHub/spotify/spotifyAPI.py [hunk 4]
```diff
@@ -48,35 +51,56 @@ def get_spotify_artist(artist_name):
...
-    response = requests.get(url, headers=headers, params=params)
+    response = requests.get(url, headers=spotify_authenticate(), params=params)
     artists_results = json.loads(response.text)
 
     for artist in artists_results['artists']['items']:
         if artist['name'] == artist_name:
             return artist['id']
 
+def get_artist_albums(artist_name):
+    album_dict = {'images':[],
```
Comment (bbelderbos) [inline webCommander/centralHub/spotify/spotifyAPI.py]: you can use `collections.defaultdict` here, then you don't have to initialize the empty lists

File: webCommander/centralHub/spotify/spotifyAPI.py [hunk 5]
```diff
@@ -48,35 +51,56 @@ def get_spotify_artist(artist_name):
...
+        'name':[],
+        'total_tracks':[]}
+
+    if len(artist_name) != SPOTIFY_FIXED_ARTIST_CHAR_LEN:
```
Comment (bbelderbos) [inline webCommander/centralHub/spotify/spotifyAPI.py]: same as above, I think we should be able to omit this if

File: webCommander/centralHub/xTwitter/twitterAPI.py [hunk 6]
```diff
@@ -62,12 +65,13 @@ def x_userlookup(x_user):
     api_url = X_GET_USER + x_user
 
     response = request_to_x_api(api_url, headers={}, params=params)
+    data_response = json.loads(response.text)
 
     # TODO: what is the right code?
-    if response.status_code != 200:
+    if 'errors' in data_response.keys():
```
Comment (bbelderbos) [inline webCommander/centralHub/xTwitter/twitterAPI.py:71]: wow so even an error condition would return a 200? strange :)

**Review by bbelderbos — CHANGES_REQUESTED** [review]
Great work, I think decoupling of get_ and search_ and getting rid of the if len() check would be better. If not possible let me know, as well as if you have other questions or stoppers ...

File: webCommander/centralHub/spotify/spotifyAPI.py (hunk 2 above)
Comment (abaddonpuff) [reply webCommander/centralHub/spotify/spotifyAPI.py]: That is right! will take it off.

File: webCommander/centralHub/models.py (hunk 1 above)
Comment (abaddonpuff) [reply webCommander/centralHub/models.py:32]: done, just need to populate.

File: webCommander/centralHub/spotify/spotifyAPI.py (hunk 2 above)
Comment (abaddonpuff) [reply webCommander/centralHub/spotify/spotifyAPI.py]: done

File: webCommander/centralHub/spotify/spotifyAPI.py (hunk 3 above)
Comment (abaddonpuff) [reply webCommander/centralHub/spotify/spotifyAPI.py]: Probably would first put the text field and then get the dropdown based on the options.

File: webCommander/centralHub/models.py [hunk 7]
```diff
@@ -32,12 +32,26 @@ def __str__(self):
...
+    def __str__(self):
+        return self.spotify_artist
+
+class SpotifyAlbumTracking(models.Model):
+    date_added = models.DateTimeField(auto_now_add=True)
+    spotify_user = models.ForeignKey(SpotifyArtistInfo,
+        related_name='artist_album',
+        on_delete=models.CASCADE)
     spotify_albums = models.CharField(max_length = 100)
     spotify_tracks = models.PositiveIntegerField(blank=True)
```
Comment (bbelderbos) [inline webCommander/centralHub/models.py:50]: `number_of_tracks`

File: webCommander/centralHub/templates/spotifyFramework/spotify_summary.html [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+          <th>Popularity</th>
+        </tr>
+    </thead>
+    <tbody>
+    {% for artist in allartists %}
+    <tr>
+        <td><a href="{% url 'show_all_artists' artist.twitter_handle %}">
+        {{ artist.spotify_artist}}
+        </a></td>
+        <br>
```
Comment (bbelderbos) [inline webCommander/centralHub/templates/spotifyFramework/spotify_summary.html:27]: remove br

File: webCommander/centralHub/spotify/spotifyAPI.py [hunk 9]
```diff
@@ -75,23 +73,28 @@ def search_spotify_artist(artist_name):
...
-        artist_id = search_spotify_artist(artist_name)
+    artist_id = search_spotify_artist(artist_name)
 
     url = SPOTIFY_BASE_URL + SPOTIFY_ARTISTS + artist_id + '/albums'
 
     response = requests.get(url, headers=spotify_authenticate())
     albums_results = json.loads(response.text)
 
+    #TODO
+    #not sure if this is the best data type for this purpose
```
Comment (bbelderbos) [inline webCommander/centralHub/spotify/spotifyAPI.py:97]: how do you mean?

File: webCommander/centralHub/spotify/spotifyAPI.py [hunk 10]
```diff
@@ -75,23 +73,28 @@ def search_spotify_artist(artist_name):
...
 
 def get_artist_albums(artist_name):
-    album_dict = {'images':[],
-        'name':[],
-        'total_tracks':[]}
+    album_dict = defaultdict(list)
 
-    if len(artist_name) != SPOTIFY_FIXED_ARTIST_CHAR_LEN:
-        artist_id = search_spotify_artist(artist_name)
+    artist_id = search_spotify_artist(artist_name)
```
Comment (bbelderbos) [inline webCommander/centralHub/spotify/spotifyAPI.py:89]: better to move out of this function, decouple the two

File: webCommander/centralHub/spotify/spotifyAPI.py [hunk 11]
```diff
@@ -75,23 +73,28 @@ def search_spotify_artist(artist_name):
     response = requests.get(url, headers=spotify_authenticate(), params=params)
     artists_results = json.loads(response.text)
 
+    #TODO Return a list of artists for a dropdown menu
```
Comment (bbelderbos) [inline webCommander/centralHub/spotify/spotifyAPI.py:76]: this you need help with right? can you detail what you try to achieve here? maybe we need to break it down into multiple steps?

**Comment by bbelderbos** [comment]
This PR is still open, can it be merged?

**Pull request closed without merging**